lib/ansible/modules/windows/win_copy.ps1

```diff
@@ -1,94 +1,141 @@
+
+# used in query and remote mode
 $force = Get-AnsibleParam -obj $params -name "force" -type "bool" -default $true
-$original_basename = Get-AnsibleParam -obj $params -name "original_basename" -type "str"
 
-# original_basename gets set if src and dest are dirs
-# but includes subdir if the source folder contains sub folders
-# e.g. you could get subdir/foo.txt
+# used in query mode, contains the local files/directories/symlinks that are to be copied
+$files = Get-AnsibleParam -obj $params -name "files" -type "list"
+$directories = Get-AnsibleParam -obj $params -name "directories" -type "list"
+$symlinks = Get-AnsibleParam -obj $params -name "symlinks" -type "list"
 
 $result = @{
     changed = $false
-    dest = $dest
-    original_basename = $original_basename
-    src = $src
 }
 
-if (($force -eq $false) -and (Test-Path -Path $dest)) {
-    $result.msg = "file already exists"
-    Exit-Json $result
+if ($diff_mode) {
+    $result.diff = @{}
 }
 
-Function Copy-Folder($src, $dest) {
-    if (Test-Path -Path $dest) {
-        if (-not (Get-Item -Path $dest -Force).PSIsContainer) {
-            Fail-Json $result "If src is a folder, dest must also be a folder. src: $src, dest: $dest"
+Function Copy-File($source, $dest) {
+    $diff = ""
+    $copy_file = $false
+    $source_checksum = $null
+    if ($force) {
+        $source_checksum = Get-FileChecksum -path $source
+    }
+
+    if (Test-Path -Path $dest -PathType Container) {
+        Fail-Json -obj $result -message "cannot copy file from $source to $($dest): dest is already a folder"
+    } elseif (Test-Path -Path $dest -PathType Leaf) {
+        if ($force) {
+            $target_checksum = Get-FileChecksum -path $dest
+            if ($source_checksum -ne $target_checksum) {
+                $copy_file = $true
+            }
         }
     } else {
-        try {
-            New-Item -Path $dest -ItemType Directory -Force -WhatIf:$check_mode
-            $result.changed = $true
-        } catch {
-            Fail-Json $result "Failed to create new folder $dest $($_.Exception.Message)"
-        }
+        $copy_file = $true
     }
 
-    foreach ($item in Get-ChildItem -Path $src) {
-        $dest_path = Join-Path -Path $dest -ChildPath $item.PSChildName
-        if ($item.PSIsContainer) {
-            Copy-Folder -src $item.FullName -dest $dest_path
-        } else {
-            Copy-File -src $item.FullName -dest $dest_path
+    if ($copy_file) {
+        $file_dir = [System.IO.Path]::GetDirectoryName($dest)
+        # validate the parent dir is not a file and that it exists
+        if (Test-Path -Path $file_dir -PathType Leaf) {
+            Fail-Json -obj $result -message "cannot copy file from $source to $($dest): object at dest parent dir is not a folder"
+        } elseif (-not (Test-Path -Path $file_dir)) {
+            # directory doesn't exist, need to create
+            New-Item -Path $file_dir -ItemType Directory -WhatIf:$check_mode | Out-Null
```

Comment:
Weird, that hasn't been the case in the past IIRC. One of the issues is how would it translate the `mode` parameter to directories. So a likely security issue when done. The other problem is that if it would fail for whatever reason (i.e. filesystem full) the directory might have been created, the action failed but there was still a change. Those are just a few reasons why in the past we expected the directory to exist.